### PR TITLE
Issue/4683 force state

### LINKED ISSUE #4683: Allow to force any state in expert mode
The backend has an api to force the service into any state. When expert mode is enabled it should be possible to select any state and force that.

Image

For example add an "Expert mode actions" column

See https://github.com/inmanta/inmanta-lsm/issues/1106

This requires an explicit confirmation with a big warning. For example:

Forcing a state might corrupt the service and the orchestrated resources. Are you sure you want to continue?

No default action, you have to choose

Image

Typing the name is not required.

### PR DESCRIPTION
**Pull request by matborowczyk**
# Description

* Short description here *

closes #4683 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Code is clear and sufficiently documented
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

### CODE REVIEW

**Comment by matborowczyk** [comment]
ci closed previous PR as they were attached to the export-bar and I missed it

File: src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+  confirmationText: string;
+  id: string;
+  targetState: string;
+  onSetInstanceState: (targetState: string) => Promise<void>;
+  isModalOpen: boolean;
+  setIsModalOpen: (isOpen: boolean) => void;
+  setErrorMessage: (message: string) => void;
+}
+
+const ConfirmationModal: React.FC<ModalProps> = ({
```
Comment (LauraStordeur) [inline src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx:134]: Wasn't there already a modal component that we could reuse ?

File: src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  );
+};
+
+interface ToastAlertMessageProps {
+  stateErrorMessage: string;
+  id: string;
+  setStateErrorMessage: React.Dispatch<React.SetStateAction<string>>;
+}
+
+const ToastAlertMessage: React.FC<ToastAlertMessageProps> = ({
```
Comment (LauraStordeur) [inline src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx:193]: Same for this one, wasn't there already a component ?

File: src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx (hunk 1 above)
Comment (matborowczyk) [reply src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx]: fixed

File: src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx (hunk 2 above)
Comment (matborowczyk) [reply src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx]: fixed

File: src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import { Maybe, VersionedServiceInstanceIdentifier } from "@/Core";
+import { ActionDisabledTooltip } from "@/UI/Components";
+import { DependencyContext } from "@/UI/Dependency";
+import { words } from "@/UI/words";
+import ConfirmationModal from "./ConfirmationModal";
+import { ToastAlertMessage } from "./ToastAlertMessage";
+
+interface Props extends VersionedServiceInstanceIdentifier {
+  instance_identity: string;
+  targets: string[];
```
Comment (LauraStordeur) [inline src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx:19]: What are the targets actually ?

File: src/UI/words.tsx [hunk 4]
```diff
@@ -96,13 +96,20 @@ const dict = {
...
   "inventory.statustab.version": "Version",
   "inventory.statustab.details": "Instance Details",
   "inventory.statustab.setInstanceState": "Set state to",
   "inventory.statustab.confirmTitle": "Confirm set state transfer",
   "inventory.statustab.confirmMessage": (name: string, state: string) =>
     `Are you sure you want to set state of instance ${name} to ${state}?`,
+  "inventory.statustab.forceState": "Force state to",
+  "inventory.statustab.forceState.confirmTitle": "Confirm force state transfer",
+  "inventory.statustab.forceState.message": (name: string, state: string) =>
+    `Are you sure you want to force state of instance ${name} to ${state}? `,
```
Comment (LauraStordeur) [inline src/UI/words.tsx:110]: Are you sure you want to force the state of instance

File: src/Slices/ServiceInventory/UI/InventoryTable.test.tsx [hunk 5]
```diff
@@ -152,16 +219,8 @@ test("ServiceInventory shows sorting buttons for sortable columns", async () =>
...
-    <MemoryRouter>
-      <InventoryTable
-        rows={[Row.a]}
-        tablePresenter={tablePresenter}
-        setSort={(v) => (sort = v)}
-        sort={{ name: "created_at", order: "desc" }}
-      />
-    </MemoryRouter>
-  );
+  const component = setup((v) => (sort = v));
```
[the remark this answers is not in the document]
Comment (matborowczyk) [reply src/Slices/ServiceInventory/UI/InventoryTable.test.tsx:222]: v for value, changed the name

File: src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx (hunk 3 above)
Comment (matborowczyk) [reply src/Slices/ServiceInventory/UI/Components/ForceStateAction.tsx:19]: possible instance  states, I'll change name as it not self-explanatory

File: src/Data/Managers/TriggerForceState/CommandManager.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      (targetState) =>
+        apiHelper.postWithoutResponse(
+          `/lsm/v1/service_inventory/${service_entity}/${id}/expert/state`,
+          environment,
+          composeCommandBody(authHelper.getUsername(), targetState, version)
+        )
+  );
+}
+
+export const composeCommandBody = (
```
[the remark this answers is not in the document]
Comment (matborowczyk) [reply src/Data/Managers/TriggerForceState/CommandManager.ts]: No worries, changed it back

**Comment by matborowczyk** [comment]
I rerun jenkins omitting e2e as was mentioned on stand-up, results from previous run are shown below
Screenshot 2023-03-29 at 15 47 52

**Review by LauraStordeur — APPROVED** [review]

**Pull request closed without merging**